### PR TITLE
Add plugable handler for `CREATE FUNCTION`

### LINKED ISSUE #9332: Plugable handler for `CREATE FUNCTION`
### Is your feature request related to a problem or challenge?

At the moment datafusion does not support `CREATE FUNCTION`, it only exposes programatic api for UDF declaration. 
Programatic approach does not support declarative `UDF`, and would not work with `SQL` only interaction.
 
I believe that we could expose a pluggable interface like `FunctionFactory`  (part of `SessionState`) where user can plug their own  `CREATE FUNCTION` handler, and plug their own declarative `UDFs`, without bringing them to core library.

### Describe the solution you'd like

A solution would allow users to declare functions like:

```sql
CREATE FUNCTION iris(FLOAT[])
RETURNS FLOAT[]
LANGUAGE TORCH
AS 'models:/iris@champion'
```

or 

```sql
CREATE FUNCTION add(BIGINT, BIGINT)
RETURNS BIGINT
LANGUAGE SQL
RETURN $1 + $2
```

`CREATE FUNCTION` handler would be configured at the `SessionContext` configuration phase:

```rust
let runtime_config = RuntimeConfig::new();
let runtime_environment = RuntimeEnv::new(runtime_config).unwrap();
let session_config =
    SessionConfig::new().set_str("datafusion.sql_parser.dialect", "PostgreSQL");
let state = SessionState::new_with_config_rt(session_config, Arc::new(runtime_environment))
    // we configure a torch function factory
    // to handle CREATE (and DROP) FUNCTION requests
    .with_function_factory(Arc::new(TorchFunctionFactory {}));

let ctx = SessionContext::new_with_state(state);
```

in this example we define `TorchFunctionFactory`,

full example can be found at https://github.com/milenkovicm/torchfusion

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by milenkovicm**
## Which issue does this PR close?

Closes #9332.

## Rationale for this change

explained in #9332

## What changes are included in this PR?

this is minimum viable product, with some open question which i hope to resolve in the process of PR review (note TODOs)

## Are these changes tested?

There is a test in place but would add more if community aggress to accept the change.
Usage example can be found at https://github.com/milenkovicm/torchfusion

## Are there any user-facing changes?

There are using facing changes, but they are backward compatible. Would update docs if community aggress to accept the change. 

A problem I see is from perspective of user experience, as generic parser does not support `CREATE FUNCTION` statement, user should change parser configuration to `PostgreSQL`. Maybe it would make sense after this change is merged adding `CREATE FUNCTION` to generic flavour

### CODE REVIEW

**Comment by milenkovicm** [comment]
Drop function will be added as well

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:499]: Since CreateFunction / DropFunction modify the catalog, what do you think about making them part of `DdlStatement` instead of a `Statement`?

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:507]: Minor: I think it would be better to follow the model of `create_view` etc and make this code a function on `SessionContext` (so that this code just dispatches to that method)

The upside of doing so is that it follows the existing code pattern and it offers a place to document `create_function` and `drop_function` publically in in more detail

File: datafusion/expr/src/logical_plan/statement.rs
Comment (alamb) [inline datafusion/expr/src/logical_plan/statement.rs:169]: I think that this approach (redefine the types) is more consistent with the rest of the DataFusion codebase, though it is an unfortunate amount of replication

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:1300]: It seems like `create` here handles two things:
1. Creation of the function
2. Registering the function with the SessionState

What if we made `create` only responsible for creating a function and made the caller responsible for registering it? I think that would make the API simpler / not involve  RW lock

For example, something like
```rust
/// A function defined by a FunctionFactory
enum NewFunction  {
  Scalar(ScalarUDF),
  Aggregate(AggregateUDF),
  Window(WindowUDF)
}

pub trait FunctionFactory: Sync + Send {
...
async fn create(
        &self,
        statement: CreateFunction,
) -> Result<NewFunction>;
...
}
```




Another alternative might be rather than passing the `SessionState`, maybe we could pass a reference to `&mut dyn FunctionRegistry` which would also avoid the RwLock 🤔

File: datafusion/core/tests/sql/sql_api.rs
Comment (alamb) [inline datafusion/core/tests/sql/sql_api.rs:99]: 🤔  maybe we should pass the config into `create` rather than `SessionContext`

File: datafusion/core/tests/sql/sql_api.rs
Comment (alamb) [inline datafusion/core/tests/sql/sql_api.rs:55]: This is very cool -- I think it would be great (as a follow on PR) if we turned this into an example (with docs, etc like the others in https://github.com/apache/arrow-datafusion/tree/main/datafusion-examples/examples)

File: datafusion/core/tests/sql/sql_api.rs
Comment (alamb) [inline datafusion/core/tests/sql/sql_api.rs:118]: Perhaps we could put this test along side the other user defined tests in https://github.com/apache/arrow-datafusion/tree/main/datafusion/core/tests/user_defined so it was easier to find

**Review by alamb — COMMENTED** [review]
Thank you for this PR @milenkovicm -- I think the idea and the PR is really neat and the code was easy to follow. ❤️ 

 I left some API suggestions, and comments in regards to your questions. Let me know what you think.

File: datafusion/core/src/execution/context/mod.rs
Comment (milenkovicm) [reply datafusion/core/src/execution/context/mod.rs:1300]: Should we want to implement sql user defined functions, we would need extension to cache logical expressions and provide them to logical analyser so it can replace udf call with the expression.

Id argue we need state or configuration as a parameter as well. Let me do other changes, as they are straightforward and we can try to find most ergonomic function

**Comment by milenkovicm** [comment]
Comments totally make sense thanks @alamb, 
it would involve a bit of creativity to nail the trait interface, let's try. I'm a bit  tied up over next few days will follow up over the weekend probably

**Comment by milenkovicm** [comment]
Items to follow up (new issues or discussion)

- [ ] Create an example and documents?
- [ ] Should `Generic` dialect support `CREATE STATEMENT`?
- [ ] Add unregister function in `SessionContext`?

File: datafusion/core/tests/sql/sql_api.rs
Comment (milenkovicm) [reply datafusion/core/tests/sql/sql_api.rs:99]: This comment is connected to previous, as getting `config` involves holding read lock. Read may not be as bad as write lock but still it will block writers. 

As `create` may take some time (like getting a ML model from repo) read lock may be problematic as well.

Function signature definitely looks better 

```rust
    async fn create(
        &self,
        state: &SessionConfig,
        statement: CreateFunction,
    ) -> Result<RegisterFunction>;
```

but lock is still an issue.

File: datafusion/core/src/execution/context/mod.rs
Comment (milenkovicm) [reply datafusion/core/src/execution/context/mod.rs:1300]: I definitely like this approach, but it has lock problem if we want to provide a `config` parameter. 

Why would we need `config`? `FunctionFactory` may store configuration for model repository url, or similar

Comment (alamb) [reply datafusion/core/src/execution/context/mod.rs]: I think you are right. I mentioned `config` because I think there was a comment about it before.

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:800]: 👍

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:1304]: I agree supporting create function would be good in the generic dialect. I made this change in https://github.com/sqlparser-rs/sqlparser-rs/pull/1159 (and I plan a sqlparser release in the next few days)

File: datafusion/core/src/execution/context/mod.rs
Comment (milenkovicm) [reply datafusion/core/src/execution/context/mod.rs]: Sorry for confusion @alamb, I believe i fix it with right amount of .clone() on right places :).
Please have a look now, I believe it looks all right. I still do not like `remove(...)` could be implemented with another enum ... have o think about it

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:842]: I think we do have them (added relatively recently): https://github.com/apache/arrow-datafusion/blob/ea30b93194cfdfe4148a45fc0e33549884ba81b1/datafusion/core/src/execution/context/mod.rs#L853-L865

File: datafusion/core/tests/user_defined/user_defined_scalar_functions.rs
Comment (alamb) [inline datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:681]: Thinking about it more, maybe we don't even need a `remove` in the `FunctionFactory` at all  -- dropping the function is mostly handled by `deregister_udf`?

Or maybe there is some sort of deiniitialization of the function that is needed that would be `FunctionFactory` specific🤔. However, perhaps UDFs that needed that could implement some sort of deregistering / deallocation on `drop`.

Handling deregistering on `drop` might be better anyways as would prevents a function from being deregistered out from under a running execution...

File: datafusion/core/src/execution/context/mod.rs
Comment (milenkovicm) [reply datafusion/core/src/execution/context/mod.rs:1304]: looking at `FunctionRegistry.remove` method, in order to implement it `FunctionRegistry` should be changed. Maybe that change is a bit too much for this PR. So i propose to remove `remove` from `FunctionFactory`  and add it back when we implement `unregister_ud*` methods in `FunctionRegistry`.

wdyt?

**Review by alamb — COMMENTED** [review]
Looking very 👨‍🍳 👌  nice @milenkovicm

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [reply datafusion/core/src/execution/context/mod.rs:1304]: I think we added the unregister_* functions recently (sorry our reviews crossed)

File: datafusion/core/tests/user_defined/user_defined_scalar_functions.rs
Comment (milenkovicm) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:681]: lets say we have a SQL UDFs which are expressed by expr, we would need to have some kind of registry (which is not udf_registry) which can be consulted by plan analyser.

so in this case we would have some kind of mock function stored in function registry (which would do nothing, just for planning) and have an expression registry, which would be used to "replace" udf's in logical plan ... 

btw, i believe adding sql udf would be trivial once we get this change and fix few issues with expression namings

Comment (alamb) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:681]: > lets say we have a SQL UDFs which are expressed by expr, we would need to have some kind of registry (which is not udf_registry) which can be consulted by plan analyser.

Another potential approach would be to encapsulate the code directly into the ScalarUDF and then use an API like `simplify()` that @jayzhan211  and I are working on in https://github.com/apache/arrow-datafusion/pull/9304 to replace the `ScalarUDF` with its `Expr` equivelnt...

Comment (milenkovicm) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:681]: will have to follow up on #9304 ...  `CREATE FUNCTION` could create UDF or UDaF ... do you maybe have idea how could we decide which unregister_ud*f to call when we get `DROP FUNCTION`?

Comment (alamb) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:681]: > do you maybe have idea how could we decide which unregister_ud*f to call when we get DROP FUNCTION?

That is a (very) good question. I don't have a good answer at the moment. I'll try and think of one

Comment (alamb) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:697]: I thought more about this last night and one potential solution is to simply try and deregister all functions with the specified name (udf, udaf, udwf, etc)

It seems non ideal however as it would allow dropping any udf function, not just the ones created by `CREATE FUNCTION` 🤔

Comment (milenkovicm) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:697]: I think that's good solution. 

On that note there can be UDF and udaf with same name, if used in aggregation query, analyser will freak out, as it will find UDF function. 

Will remove `remove` function from trait for now, we can always put it back

Comment (milenkovicm) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs]: `remove` has been gone :), `DROP FUNCTION` will not discriminate against udf type and drop them all.

`remove` is dropped with condition that `simplify` from #9304 find its way to other function types, otherwise function factory would have to implement complicated registry with weak ref or something like that

File: datafusion/core/src/execution/context/mod.rs
Comment (milenkovicm) [inline datafusion/core/src/execution/context/mod.rs:1320]: The last thing standing is trait name, I'm not sure if `FunctionFactory` is the best name for it. It would be appropriate if this PR is part of spark framework :) 

 any suggestions @alamb ?

File: datafusion/core/tests/user_defined/user_defined_scalar_functions.rs
Comment (alamb) [inline datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:735]: Would it also be possible to add a test using `SessionContext::sql_with_options` showing that if DDL is disabled that CREATE / DROP function is also disabled?


https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionContext.html#method.sql_with_options

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (alamb) [inline datafusion/expr/src/logical_plan/ddl.rs:326]:
```suggestion
/// Arguments passed to `CREATE FUNCTION`
///
/// Note this meant to be the same as from sqlparser's [`sqlparser::ast::Statement::CreateFunction`]
#[derive(Clone, PartialEq, Eq, Hash, Debug)]
pub struct CreateFunction {
```

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (alamb) [inline datafusion/expr/src/logical_plan/ddl.rs:327]: Is this still TODO (or are the rest of the comments below todo)? I think this structure makes sense and with a little more explanation on rationale (aka the Statement::CreateFunction is a variant and not its own struct) this make sense

The rationale I think would be important so that when we upgrade sqlparser's version and it adds more CREATE FUNCTION variants, the updater knows the desired results in DataFusion

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (alamb) [inline datafusion/expr/src/logical_plan/ddl.rs:344]: I think it is fine to leave some of these fields as "TODO" items and if someone needs them they can add them as a follow on PR
```suggestion
    // TODO: figure out how to support mode
```

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (alamb) [inline datafusion/expr/src/logical_plan/ddl.rs:385]: Does this really need a schema?

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (alamb) [inline datafusion/expr/src/logical_plan/ddl.rs:339]: Why is this needed?

File: datafusion/sql/src/statement.rs
Comment (alamb) [inline datafusion/sql/src/statement.rs:680]: I recommend in this case that we throw an error if someone tries to create a qualified function, and if it is important to support qualified function names in the future we can revisit the issue then

So
```
CREATE FUNCTION foo
```
would work

but
```
CREATE FUNCTION my.foo ... 
```

would error (not yet implemented error)



This is the corresponding code for table references https://github.com/apache/arrow-datafusion/blob/dd4263f843e093c807d63edf73a571b1ba2669b5/datafusion/common/src/column.rs#L71-L99

 is https://docs.rs/datafusion/latest/datafusion/sql/enum.TableReference.html 

But at the moment DataFusion doesn't put functions in a schema or catalog, it simply puts them some global namespace.

File: datafusion/sql/src/statement.rs
Comment (alamb) [inline datafusion/sql/src/statement.rs:728]: See above -- I also recommend erroring in the case of multi part identifiers.

**Review by alamb — COMMENTED** [review]
Thank you @milenkovicm  -- I think this PR is almost good to go from my perspective. I had a few minor comments, but nothing major.

I think we should have at least one other maintainer give it a look. Perhaps @comphead or @universalmind303 ?

One interesting implication of this feature is that once we migrate all functions to be "udfs" (e.g. https://github.com/apache/arrow-datafusion/issues/8045) it means we can now drop functions that weren't explicitly made with `CREATE FUNCTION`

For example, on this branch I can drop the `abs` function:

```sql
DataFusion CLI v36.0.0
❯ DROP FUNCTION abs;
0 rows in set. Query took 0.009 seconds.
```

Although I do get an appropriate error when I try to drop a function that is still a `BuiltInScalarFunction` like `add`:
```sql
❯ DROP FUNCTION add;
Execution error: Function does not exist
```

People could

File: datafusion/core/tests/user_defined/user_defined_scalar_functions.rs
Comment (milenkovicm) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs:718]: Is this a comment to be added?

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (milenkovicm) [reply datafusion/expr/src/logical_plan/ddl.rs:342]: Its not needed

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (milenkovicm) [reply datafusion/expr/src/logical_plan/ddl.rs:387]: `DdlStatement::schema` needs it. I've tried to remove it and create it when needed but compiler complains it does not live long enough

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (milenkovicm) [reply datafusion/expr/src/logical_plan/ddl.rs:342]: If you refer to `schema` same comment like above

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (milenkovicm) [reply datafusion/expr/src/logical_plan/ddl.rs:387]: Has to be `pub`. Other `structs` from that package have same property

File: datafusion/core/tests/user_defined/user_defined_scalar_functions.rs
Comment (alamb) [reply datafusion/core/tests/user_defined/user_defined_scalar_functions.rs]: Oh, yes, sorry -- I meant it to be a suggestion for a comment. Sorry about that

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [reply datafusion/core/src/execution/context/mod.rs:1320]: I think the name is good and describes what this trait does -- namely it creates functions. I can't think of anything better or more appropriate

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:1387]: BTW once I release the next version of sqlparser-rs (hopefully today) this will no longer be true

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (alamb) [reply datafusion/expr/src/logical_plan/ddl.rs:340]: yes, I meant I think we can remove the schema here

File: datafusion/expr/src/logical_plan/ddl.rs
Comment (alamb) [reply datafusion/expr/src/logical_plan/ddl.rs:385]: Got it -- I tried locally and i see that `DdlStatement::schema()` returns a reference. Makes sense

**Review by alamb — APPROVED** [review]
Thanks again @milenkovicm  -- the only thing left from my perspective is to "error on multi-part identifiers" which I think we could do as a follow on PR if needed. 

However, I do think we should leave this open for a few more days to let other people have a chance to comment on it before we merge.

Thanks again. This is a *really* neat feature that I think will let people build all sorts of cool things using DataFusion 🙏

File: datafusion/core/src/execution/context/mod.rs
Comment (milenkovicm) [reply datafusion/core/src/execution/context/mod.rs:1387]: updated

**Comment by milenkovicm** [comment]
> Thanks again @milenkovicm -- the only thing left from my perspective is to "error on multi-part identifiers" which I think we could do as a follow on PR if needed.
> 
> However, I do think we should leave this open for a few more days to let other people have a chance to comment on it before we merge.
> 
> Thanks again. This is a _really_ neat feature that I think will let people build all sorts of cool things using DataFusion 🙏

@alamb  multi-part identifiers fixed as well, please have a look.

also, can you please let me know once new version of `sqlparser` lands, will remove dialect configuration from test.

thanks a lot for all your help

**Comment by milenkovicm** [comment]
and please squash commit on merge, thanks

**Comment by alamb** [comment]
> and please squash commit on merge, thanks

Yes, this will be done (as on all commits)

**Review by alamb — APPROVED** [review]
Thanks again @milenkovicm

File: datafusion/core/tests/sql/sql_api.rs
Comment (comphead) [inline datafusion/core/tests/sql/sql_api.rs:21]: is it related?

File: datafusion/core/src/execution/context/mod.rs
Comment (comphead) [inline datafusion/core/src/execution/context/mod.rs:843]: please use `exec_err!` macros

File: datafusion/core/tests/sql/sql_api.rs
Comment (comphead) [reply datafusion/core/tests/sql/sql_api.rs]: Right, this functionality looks important and definitely needs to be covered with docs, examples in following PR, looking forward for it

**Review by comphead — COMMENTED** [review]
Thanks @milenkovicm 
I would say it is nice to get also tests what will happen if the function is invalid/not compilable or something

another question if function can call another function?

**Comment by milenkovicm** [comment]
> Thanks @milenkovicm I would say it is nice to get also tests what will happen if the function is invalid/not compilable or something
> 
> another question if function can call another function?

Thanks for your comments @comphead , 
can you please elaborate your comment, thanks

**Comment by comphead** [comment]
> > Thanks @milenkovicm I would say it is nice to get also tests what will happen if the function is invalid/not compilable or something
> > another question if function can call another function?
> 
> Thanks for your comments @comphead , can you please elaborate your comment, thanks

Its great to have a test with invalid function, the function which is not parseable or compilable.

And the next question if the user created functions `f1` and `f2`, can `f1` be called inside `f2`?

**Comment by milenkovicm** [comment]
> > > Thanks @milenkovicm I would say it is nice to get also tests what will happen if the function is invalid/not compilable or something
> > > another question if function can call another function?
> > 
> > 
> > Thanks for your comments @comphead , can you please elaborate your comment, thanks
> 
> Its great to have a test with invalid function, the function which is not parseable or compilable.

could be added, no problem

> And the next question if the user created functions `f1` and `f2`, can `f1` be called inside `f2`?

I see two cases here:
1. if we have functions defined in rust, there is no change to current behaviour I believe
2. if we have `SQL UDF` then analyser should replace f1 first, then f2, so it should run as well

do i miss something here?

**Comment by alamb** [comment]
> Its great to have a test with invalid function, the function which is not parseable or compilable.
> 
> And the next question if the user created functions `f1` and `f2`, can `f1` be called inside `f2`?

I think it is important to note that this PR does NOT actually add any support for `CREATE FUNCTION` by default in DataFusion. Instead it sets up the plumbing for people to add their own `CREATE FUNCTION` support

Thus, the answer to this question is "it depends on whatever the user has coded the `FunctionFactory` to do

**Comment by alamb** [comment]
I took the liberty of pushing a commit with a sqllogictest that shows the behavior of `CREATE FUNCTION` / `DROP FUNCTION` which should hopefully make this PR a little clearer

**Comment by alamb** [comment]
> also, can you please let me know once new version of sqlparser lands, will remove dialect configuration from test.

I just released sqlparser https://crates.io/crates/sqlparser/0.44.0 (but we will need to update DataFusion to use it first)

File: datafusion/core/src/execution/context/mod.rs
Comment (jayzhan211) [inline datafusion/core/src/execution/context/mod.rs:842]: It is better to add the `if exist` test like `DROP FUNCTION if exists abs;` so the logic here is covered.

**Comment by milenkovicm** [comment]
Thanks a lot @alamb 

As follow up, maybe we should consider modifying create function syntax, to add support for named parameters and global variables ( if that's correct name `@var` and `@@var` )

Then next step would be adding `sql function contrib` 😀 similar to databrick's.

File: datafusion/core/src/execution/context/mod.rs
Comment (alamb) [reply datafusion/core/src/execution/context/mod.rs:842]: That is a good call -- I added a test to sqllogictest

**Comment by alamb** [comment]
I plan to leave this PR open for another day or two to give people a chance to comment if they would like to and then I'll merge it in.

Thanks again @milenkovicm

**Comment by milenkovicm** [comment]
Thanks for all your help @alamb!

Please let me know when new sqlparser lands so I can update dialect used in test

**Comment by alamb** [comment]
> Please let me know when new sqlparser lands so I can update dialect used in test

Thank you -- I saw https://github.com/apache/arrow-datafusion/pull/9447 from dependabot and I think @JasonLi-cn  has implemented the upgrade as part of https://github.com/apache/arrow-datafusion/pull/9268

I will attempt to review https://github.com/apache/arrow-datafusion/pull/9268 later today

**Comment by alamb** [comment]
https://github.com/apache/arrow-datafusion/pull/9268 was merged, so the new sqlparser release should now be available.

**Comment by alamb** [comment]
I merged up from main to pick up a fix for the CI

**Comment by milenkovicm** [comment]
Thanks @alamb

**Comment by alamb** [comment]
Merged up from main again to resolve conflicts

**Comment by alamb** [comment]
Thanks again @milenkovicm  -- this is a pretty epic feature

**Pull request merged**
